gnosis/cowswap
# [Error handling] Remodel actions

Continuation of https://github.com/gnosis/cowswap/pull/805 (related to https://github.com/gnosis/cowswap/issues/782)

Model the actions closer to real UI actions:
* `getNewQuoteStart`:  Signal that now is fetching a new price, invalidating any previous price and activating the loader
* `refreshQuoteStart`: Signal that now is refreshing the current price, activating the loader
* `updateQuote`: We have a new price, this actions updates it and deactivate the loader
* `setQuoteError`: There was an error getting the new price or refreshing, this actions updates it and deactivate the loader


## Next PR
* Handle the offline case to be more specific when an error occurs

## Test instructions
* Try to reproduce not enough liquidity error
* Try to reproduce unsupported error. U can use kick token `0xc12d1c73ee7dc3615ba4e37e4abfdbddfa38907e`
* Try to reproduce a network error (dev tools helps with this), verify the error disappears automatically once the network is back
* Try to reproduce GPv2 is down (see below how to)
* Try to reproduce Paraswap is down (see below how to)
* Try to reproduce GPv2 and Paraswap is down (see below how to)
* IMPORTANT: Make sure that it works that when we have for example an unsupported token in our GPv2 we don't allow to send orders even if we get a price for paraswap. Our API informs about some issues that makes a user non-usable even if some other exchange allows to use it


Some help to reproduce errors in GPv2 or Paraswap. Just add a throw here:
Simulating an error, is also easy, u just throw an error here: 
 - [GPv2](https://github.com/gnosis/cowswap/blob/d97e6cd506ee1b6ae9117d57757fa691a6f3059f/src/custom/utils/operator/index.ts#L201)
- and here for [Paraswap](https://github.com/gnosis/cowswap/blob/d97e6cd506ee1b6ae9117d57757fa691a6f3059f/src/custom/utils/paraswap.ts#L57-L57)


## Screenshots
Unsupported tokens still works:

![image](https://user-images.githubusercontent.com/2352112/123431215-4fdda280-d5c9-11eb-9293-189f2aec81c6.png)

Message when there's no network:
![image](https://user-images.githubusercontent.com/2352112/123431742-e5793200-d5c9-11eb-8710-3d1a2a41b3cf.png)
